Handle exceptions raised from non-inlined functions

aorenste opened the pull request:
--- Closing this PR for now - there are a bunch of edge conditions which need consideration before it can be dealt with properly. In particular there are some tests that expect the existing throw behavior so fixing it breaks those tests ---

When a non-inlined function called during tracing raises an exception we need to make sure to handle it similarly to when an inline function raises one - currently that means a graph break but in the future could be jumping to a traced `except` clause.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130011



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

aorenste committed:
Fix exception raised from called C function

[ghstack-poisoned]

aorenste committed:
Update on "Fix exception raised from called C function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

aorenste commented:
@pytorchbot rebase

aorenste committed:
Update

[ghstack-poisoned]

aorenste committed:
Update on "Fix exception raised from called C function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

Skylion007 commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -751,7 +751,26 @@ def call_function(
+        except RuntimeError as e:
+            # Our function call raised an exception. Treat it as an internal
+            # exception and "handle" it.
+            log.debug("called function raised exception: %s (%s)", type(e), e)
```
use `exc_info=True` instead

aorenste committed:
Update on "Handle exceptions raised from non-inlined functions"

When a non-inlined function called during tracing raises an exception we need to make sure to handle it similarly to when an inline function raises one - currently that means a graph break but in the future could be jumping to a traced `except` clause.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]
